libraries/WebServer/src/Middlewares.h

```diff
@@ -0,0 +1,208 @@
+#ifndef MIDDLEWARES_H
+#define MIDDLEWARES_H
+
+#include <WebServer.h>
+#include <Stream.h>
+
+#include <assert.h>
+
+// curl-like logging middleware
+class LoggingMiddleware : public Middleware {
+private:
+  Stream *_out;
+
+public:
+  explicit LoggingMiddleware(Stream &out) : _out(&out) {}
+
+  bool run(WebServer &server, Middleware::Callback next) override {
+    _out->print(F("* Connection from "));
+    _out->print(server.client().remoteIP().toString());
+    _out->print(F(":"));
+    _out->println(server.client().remotePort());
+
+    _out->print(F("< "));
+    HTTPMethod method = server.method();
+    if (method == HTTP_ANY) {
+      _out->print(F("HTTP_ANY"));
+    } else {
+      _out->print(http_method_str((http_method)method));
+    }
+    _out->print(F(" "));
+    _out->print(server.uri());
+    _out->print(F(" "));
+    _out->println(server.version());
+
+    int n = server.headers();
+    for (int i = 0; i < n; i++) {
+      String v = server.header(i);
+      if (!v.isEmpty()) {
+        // because these 2 are always there, eventually empty: "Authorization", "If-None-Match"
+        _out->print(F("< "));
+        _out->print(server.headerName(i));
+        _out->print(F(": "));
+        _out->println(server.header(i));
+      }
+    }
+
+    _out->println(F("<"));
+
+    bool ret = next();
```

Comment:
similar to that case https://github.com/espressif/arduino-esp32/pull/10186/files#r1774121401